File: src/main/java/ee/carlrobert/codegpt/actions/GenerateGitCommitMessageAction.java
package ee.carlrobert.codegpt.actions;

import static com.intellij.openapi.ui.Messages.OK;

import com.intellij.notification.Notification;
import com.intellij.notification.NotificationType;
import com.intellij.notification.Notifications;
import com.intellij.openapi.actionSystem.ActionUpdateThread;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.diff.impl.patch.IdeaTextPatchBuilder;
import com.intellij.openapi.diff.impl.patch.UnifiedDiffWriter;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vcs.VcsDataKeys;
import com.intellij.openapi.vcs.VcsException;
import com.intellij.vcs.commit.CommitWorkflowUi;
import ee.carlrobert.codegpt.CodeGPTBundle;
import ee.carlrobert.codegpt.EncodingManager;
import ee.carlrobert.codegpt.Icons;
import ee.carlrobert.codegpt.completions.CommitMessageCompletionParameters;
import ee.carlrobert.codegpt.completions.CompletionRequestService;
import ee.carlrobert.codegpt.settings.prompts.CommitMessageTemplate;
import ee.carlrobert.codegpt.ui.OverlayUtil;
import ee.carlrobert.codegpt.util.CommitWorkflowChanges;
import ee.carlrobert.codegpt.util.GitUtil;
import ee.carlrobert.llm.client.openai.completion.ErrorDetails;
import ee.carlrobert.llm.completion.CompletionEventListener;
import git4idea.repo.GitRepository;
import java.io.IOException;
import java.io.StringWriter;
import java.util.Optional;
import java.util.concurrent.ExecutionException;
import okhttp3.sse.EventSource;
import org.jetbrains.annotations.NotNull;

public class GenerateGitCommitMessageAction extends AnAction {

  public static final int MAX_TOKEN_COUNT_WARNING = 8196;
  private final EncodingManager encodingManager;

  public GenerateGitCommitMessageAction() {
    super(
        CodeGPTBundle.get("action.generateCommitMessage.title"),
        CodeGPTBundle.get("action.generateCommitMessage.description"),
        Icons.DefaultSmall);
    encodingManager = EncodingManager.getInstance();
  }

  @Override
  public void update(@NotNull AnActionEvent event) {
    ApplicationManager.getApplication().invokeLater(() -> {
      var commitWorkflowUi = event.getData(VcsDataKeys.COMMIT_WORKFLOW_UI);
      if (commitWorkflowUi == null) {
        event.getPresentation().setVisible(false);
        return;
      }

      var callAllowed = CompletionRequestService.isRequestAllowed();
      event.getPresentation().setEnabled(callAllowed
          && new CommitWorkflowChanges(commitWorkflowUi).isFilesSelected());
      event.getPresentation().setText(CodeGPTBundle.get(callAllowed
          ? "action.generateCommitMessage.title"
          : "action.generateCommitMessage.missingCredentials"));
    });
  }

  @Override
  public void actionPerformed(@NotNull AnActionEvent event) {
    var project = event.getProject();
    if (project == null || project.getBasePath() == null) {
      return;
    }

    var gitDiff = getDiff(event, project);
    var tokenCount = encodingManager.countTokens(gitDiff);
    if (tokenCount > MAX_TOKEN_COUNT_WARNING
        && OverlayUtil.showTokenSoftLimitWarningDialog(tokenCount) != OK) {
      return;
    }

    var commitWorkflowUi = event.getData(VcsDataKeys.COMMIT_WORKFLOW_UI);
    if (commitWorkflowUi != null) {
      CompletionRequestService.getInstance().getCommitMessageAsync(
          new CommitMessageCompletionParameters(
              gitDiff,
              project.getService(CommitMessageTemplate.class).getSystemPrompt()),
          getEventListener(project, commitWorkflowUi));
    }
  }

  @Override
  public @NotNull ActionUpdateThread getActionUpdateThread() {
    return ActionUpdateThread.EDT;
  }

  private String getDiff(AnActionEvent event, Project project) {
    var commitWorkflowUi = Optional.ofNullable(event.getData(VcsDataKeys.COMMIT_WORKFLOW_UI))
        .orElseThrow(() -> new IllegalStateException("Could not retrieve commit workflow ui."));

    try {
      GitRepository repository = ApplicationManager.getApplication()
          .executeOnPooledThread(() -> GitUtil.getProjectRepository(project))
          .get();

      try {
        var includedChanges = commitWorkflowUi.getIncludedChanges();
        var filePatches = IdeaTextPatchBuilder.buildPatch(
            project, includedChanges, repository.getRoot().toNioPath(), false, true);
        var diffWriter = new StringWriter();
        UnifiedDiffWriter.write(
            null,
            repository.getRoot().toNioPath(),
            filePatches,
            diffWriter,
            "\n",
            null,
            null
        );
        return diffWriter.toString();
      } catch (VcsException | IOException e) {
        throw new RuntimeException("Unable to create git diff", e);
      }
    } catch (InterruptedException | ExecutionException e) {
      throw new RuntimeException(e);
    }
  }

  private CompletionEventListener<String> getEventListener(
      Project project,
      CommitWorkflowUi commitWorkflowUi) {
    return new CompletionEventListener<>() {
      private final StringBuilder messageBuilder = new StringBuilder();

      @Override
      public void onMessage(String message, EventSource eventSource) {
        messageBuilder.append(message);
        updateCommitMessage(messageBuilder.toString());
      }

      @Override
      public void onComplete(StringBuilder result) {
        if (messageBuilder.isEmpty()) {
          updateCommitMessage(result.toString());
        }
      }

      private void updateCommitMessage(String message) {
        ApplicationManager.getApplication().invokeLater(() ->
            WriteCommandAction.runWriteCommandAction(project, () ->
                commitWorkflowUi.getCommitMessageUi().setText(message)
            )
        );
      }

      @Override
      public void onError(ErrorDetails error, Throwable ex) {
        Notifications.Bus.notify(new Notification(
            "CodeGPT Notification Group",
            "CodeGPT",
            error.getMessage(),
            NotificationType.ERROR));
      }
    };
  }
}

File: src/main/kotlin/ee/carlrobert/codegpt/actions/GenerateGitCommitMessageAction.kt
package ee.carlrobert.codegpt.actions

import com.intellij.notification.Notification
import com.intellij.notification.NotificationType
import com.intellij.notification.Notifications
import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.editor.Document
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.ex.EditorEx
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.openapi.vcs.VcsDataKeys
import com.intellij.openapi.vcs.ui.CommitMessage
import com.intellij.vcs.commit.CommitWorkflowUi
import ee.carlrobert.codegpt.CodeGPTBundle
import ee.carlrobert.codegpt.EncodingManager
import ee.carlrobert.codegpt.Icons
import ee.carlrobert.codegpt.completions.CompletionRequestService
import ee.carlrobert.codegpt.completions.CompletionRequestService.isRequestAllowed
import ee.carlrobert.codegpt.settings.GeneralSettings
import ee.carlrobert.codegpt.settings.configuration.CommitMessageTemplate
import ee.carlrobert.codegpt.settings.service.ServiceType
import ee.carlrobert.codegpt.settings.service.ServiceType.YOU
import ee.carlrobert.codegpt.ui.OverlayUtil
import ee.carlrobert.llm.client.openai.completion.ErrorDetails
import ee.carlrobert.llm.completion.CompletionEventListener
import okhttp3.sse.EventSource
import java.io.BufferedReader
import java.io.File
import java.io.IOException
import java.io.InputStreamReader
import java.nio.file.Files
import java.nio.file.Path
import java.util.Objects
import java.util.Optional
import java.util.stream.Collectors.joining

class GenerateGitCommitMessageAction : AnAction(
  CodeGPTBundle.get("action.generateCommitMessage.title"),
  CodeGPTBundle.get("action.generateCommitMessage.description"),
  Icons.Sparkle) {

  private val encodingManager: EncodingManager = EncodingManager.getInstance()

  override fun update(event: AnActionEvent) {
    val commitWorkflowUi = event.getData(VcsDataKeys.COMMIT_WORKFLOW_UI)
    if (GeneralSettings.isSelected(YOU) || commitWorkflowUi == null) {
      event.presentation.isVisible = false
      return
    }

    val callAllowed = isRequestAllowed()
    event.presentation.isEnabled = callAllowed
            && CommitWorkflowChanges(commitWorkflowUi).isFilesSelected
    event.presentation.text = CodeGPTBundle.get(
      if (callAllowed)
        "action.generateCommitMessage.title"
      else "action.generateCommitMessage.missingCredentials")
  }

  override fun actionPerformed(event: AnActionEvent) {
    val project = event.project ?: return
    project.basePath ?: return

    val gitDiff = getGitDiff(event, project)
    val tokenCount = encodingManager.countTokens(gitDiff)
    if (tokenCount > MAX_TOKEN_COUNT_WARNING
      && OverlayUtil.showTokenSoftLimitWarningDialog(tokenCount) != Messages.OK) {
      return
    }

    val editor = getCommitMessageEditor(event) as? EditorEx ?: return
    editor.setCaretVisible(false)
    CompletionRequestService.getInstance().generateCommitMessageAsync(
      project.getService(CommitMessageTemplate::class.java).getSystemPrompt(),
      gitDiff,
      getEventListener(project, editor.document))
  }

  override fun getActionUpdateThread(): ActionUpdateThread {
    return ActionUpdateThread.EDT
  }

  private fun getEventListener(project: Project, document: Document): CompletionEventListener<String> {
    return object : CompletionEventListener<String> {
      private val messageBuilder = StringBuilder()

      override fun onMessage(message: String?, eventSource: EventSource) {
        messageBuilder.append(message)
        val application = ApplicationManager.getApplication()
        application.invokeLater {
          application.runWriteAction {
            WriteCommandAction.runWriteCommandAction(project) {
              document.setText(messageBuilder)
            }
          }
        }
      }

      override fun onError(error: ErrorDetails, ex: Throwable) {
        Notifications.Bus.notify(Notification(
            "CodeGPT Notification Group",
            "CodeGPT",
            error.message,
            NotificationType.ERROR))
      }
    }
  }

  private fun getCommitMessageEditor(event: AnActionEvent): Editor? {
    return (event.getData(VcsDataKeys.COMMIT_MESSAGE_CONTROL) as? CommitMessage)?.editorField?.editor
  }

  private fun getGitDiff(event: AnActionEvent, project: Project): String {
    val commitWorkflowUi = Optional.ofNullable(event.getData(VcsDataKeys.COMMIT_WORKFLOW_UI))
      .orElseThrow { IllegalStateException("Could not retrieve commit workflow ui.") }
    val changes = CommitWorkflowChanges(commitWorkflowUi)
    val projectBasePath = project.basePath
    val gitDiff = getGitDiff(projectBasePath, changes.includedVersionedFilePaths, false)
    val stagedGitDiff = getGitDiff(projectBasePath, changes.includedVersionedFilePaths, true)
    val newFilesContent = getNewFilesDiff(projectBasePath, changes.includedUnversionedFilePaths)

    return mapOf(
      "Git diff" to gitDiff,
      "Staged git diff" to stagedGitDiff,
      "New files" to newFilesContent)
      .entries.stream()
      .filter { !it.value.isNullOrBlank() }
      .map { "%s:%n%s".formatted(it.key, it.value) }
      .collect(joining("\n\n"))
  }

  private fun getGitDiff(projectPath: String?, filePaths: List<String>, cached: Boolean): String {
    if (filePaths.isEmpty()) {
      return ""
    }

    val process = createGitDiffProcess(projectPath, filePaths, cached)
    return BufferedReader(InputStreamReader(process.inputStream)).lines().collect(joining("\n"))
  }

  private fun getNewFilesDiff(projectPath: String?, filePaths: List<String>): String? {
    return filePaths.stream()
      .map { pathString ->
        val filePath = Path.of(pathString)
        val relativePath = Path.of(projectPath).relativize(filePath)
        try {
          return@map """
            New file '$relativePath' content:
            ${Files.readString(filePath)}
            """.trimIndent()
        } catch (ignored: IOException) {
          return@map null
        }
      }
      .filter(Objects::nonNull)
      .collect(joining("\n"))
  }

  private fun createGitDiffProcess(projectPath: String?, filePaths: List<String>, cached: Boolean): Process {
    val command = mutableListOf("git", "diff")
    if (cached) {
      command.add("--cached")
    }
    command.addAll(filePaths)

    val processBuilder = ProcessBuilder(command)
    processBuilder.directory(File(projectPath))
    try {
      return processBuilder.start()
    } catch (ex: IOException) {
      throw RuntimeException("Unable to start git diff process", ex)
    }
  }

  internal class CommitWorkflowChanges(commitWorkflowUi: CommitWorkflowUi) {
    val includedVersionedFilePaths: List<String> = commitWorkflowUi.getIncludedChanges().stream()
      .map { it.virtualFile }
      .filter(Objects::nonNull)
      .map { it!!.path }
      .toList()
    val includedUnversionedFilePaths: List<String> =
      commitWorkflowUi.getIncludedUnversionedFiles().stream().map { it.path }.toList()

    val isFilesSelected =
      includedVersionedFilePaths.isNotEmpty() || includedUnversionedFilePaths.isNotEmpty()
  }

  companion object {
    const val MAX_TOKEN_COUNT_WARNING: Int = 4096
  }
}
